src/dataplane-protocol/src/record.rs

```diff
@@ -20,67 +19,82 @@ use crate::core::Version;
 use crate::derive::Decode;
 use crate::derive::Encode;
 
-use crate::batch::DefaultBatch;
+use crate::batch::Batch;
 use crate::Offset;
 
-pub type DefaultRecord = Record<DefaultAsyncBuffer>;
-
 /// maximum text to display
 static MAX_STRING_DISPLAY: Lazy<usize> = Lazy::new(|| {
     let var_value = std::env::var("FLV_MAX_STRING_DISPLAY").unwrap_or_default();
     var_value.parse().unwrap_or(16384)
 });
 
-/// slice that can works in Async Context
-pub trait AsyncBuffer {
-    fn len(&self) -> usize;
+#[deprecated(since = "0.6.0", note = "Use plain Record instead")]
+pub type DefaultRecord = Record<RecordData>;
+
+/// A key for determining which partition a record should be sent to.
+///
+/// This type is used to support conversions from any other type that
+/// may be converted to a `Vec<u8>`, while still allowing the ability
+/// to explicitly state that a record may have no key (`RecordKey::NULL`).
+///
+/// # Examples
+///
+/// ```
+/// # use fluvio_dataplane_protocol::record::RecordKey;
+/// let key = RecordKey::NULL;
+/// let key: RecordKey = "Hello, world!".into();
+/// let key: RecordKey = String::from("Hello, world!").into();
+/// let key: RecordKey = vec![1, 2, 3, 4].into();
+/// ```
+pub struct RecordKey(RecordKeyInner);
+
+impl RecordKey {
+    pub const NULL: Self = Self(RecordKeyInner::Null);
+
+    fn into_option(self) -> Option<RecordData> {
+        match self.0 {
+            RecordKeyInner::Key(key) => Some(key),
+            RecordKeyInner::Null => None,
+        }
+    }
 }
 
-pub trait Records {}
-
-#[derive(Default)]
-pub struct DefaultAsyncBuffer(Bytes);
+enum RecordKeyInner {
+    Null,
+    Key(RecordData),
+}
 
-impl DefaultAsyncBuffer {
-    pub fn new<T>(val: T) -> Self
-    where
-        T: Into<Bytes>,
-    {
-        DefaultAsyncBuffer(val.into())
+impl<K: Into<Vec<u8>>> From<K> for RecordKey {
+    fn from(k: K) -> Self {
+        Self(RecordKeyInner::Key(RecordData::from(k)))
     }
+}
+
+#[deprecated(since = "0.6.0", note = "Use RecordData instead")]
+pub type DefaultAsyncBuffer = RecordData;
+
+#[derive(Default)]
+pub struct RecordData(Bytes);
 
+impl RecordData {
     pub fn len(&self) -> usize {
         self.0.len()
     }
+}
 
-    /// Check if value is binary content
-    pub fn is_binary(&self) -> bool {
```

Comment:
why is this API removed?

Reply:
From what I can tell it is completely unused

Reply:
I reverted any changes that should be breaking, so this can stay